### PR TITLE
[improvement](fe) Rewrite count over row-independent expressions

### PR DESCRIPTION
**Pull request by mrhhsg**
### What problem does this PR solve?

Issue Number: None

Problem Summary:
`count(expr)` currently still has to evaluate a row-independent expression for each input row when the expression is not represented as a literal, for example `count(json_extract('<constant json>', '$.path'))`. For a deterministic row-independent expression, the aggregate result only depends on whether that expression is NULL: it is either `0` or equivalent to `count(*)`.

This patch extends the existing `CountLiteralRewrite` rule for non-distinct `count(<row-independent expression>)`. The rule rewrites foldable, deterministic, slot-free, non-literal expressions to `if(<expr> is null, 0, count(*))`, so the row-independent expression is evaluated above aggregation instead of per input row.

There is one accepted trade-off: compared with original `count(expr)`, an empty input would not evaluate `expr`, while the rewritten shape may evaluate the row-independent expression once per output group to decide whether it is NULL. The code documents this choice. Avoiding that evaluation by injecting an aggregate dependency into scalar subtrees can break functions whose BE implementation requires specific arguments to remain constant, such as `date_trunc`'s time unit.

Runtime-sensitive or unsafe shapes such as `sleep()`, nondeterministic/volatile expressions, nested aggregates, subqueries, table-generating functions, and window expressions are not rewritten.

The patch also adds FE unit coverage and a regression test covering non-NULL constants, NULL constants, grouped aggregation, the motivating multi-path JSON expression, column-dependent negative cases, empty input, runtime-empty input, nested expressions over `count(expr)`, functions with required constant arguments, and `sleep()` exclusion.

### Release note

None

### Check List (For Author)

- Test:
    - `env -u HTTP_PROXY -u HTTPS_PROXY -u ALL_PROXY -u http_proxy -u https_proxy -u all_proxy ./run-fe-ut.sh --run org.apache.doris.nereids.rules.rewrite.CountLiteralRewriteTest`
    - `env -u HTTP_PROXY -u HTTPS_PROXY -u ALL_PROXY -u http_proxy -u https_proxy -u all_proxy ./build.sh --fe`
    - `env -u HTTP_PROXY -u HTTPS_PROXY -u ALL_PROXY -u http_proxy -u https_proxy -u all_proxy ./run-regression-test.sh --conf output/local-regression/regression-conf-46001.groovy --run -d nereids_rules_p0/count_constant_rewrite -s count_constant_rewrite -forceGenOut`
    - `env -u HTTP_PROXY -u HTTPS_PROXY -u ALL_PROXY -u http_proxy -u https_proxy -u all_proxy ./run-regression-test.sh --conf output/local-regression/regression-conf-46001.groovy --run -d nereids_rules_p0/count_constant_rewrite -s count_constant_rewrite`
    - `env -u HTTP_PROXY -u HTTPS_PROXY -u ALL_PROXY -u http_proxy -u https_proxy -u all_proxy ./run-regression-test.sh --conf output/local-regression/regression-conf-46001.groovy --run -d statistics`
    - `git diff --check`
    - `git diff --check $(git merge-base HEAD origin/master)`
- Behavior changed: Yes. Nereids can rewrite `count()` over deterministic row-independent nonliteral expressions to `if(expr is null, 0, count(*))`; the row-independent expression may be evaluated once per output group.
- Does this need documentation: No

### CODE REVIEW

**Comment by hello-stephen** [comment]
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Analyzer.java [hunk 1]
```diff
@@ -196,7 +197,10 @@ private static List<RewriteJob> buildAnalyzerJobs() {
             //       -->project(encode A as x)
             //            -->scan（T)
             bottomUp(new CompressedMaterialize()),
-            topDown(new NormalizeAggregate()),
+            topDown(
+                    new CountConstantRewrite(),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Analyzer.java]: do not add a new rule, update `org.apache.doris.nereids.rules.rewrite.CountLiteralRewrite` instead

Comment (mrhhsg) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Analyzer.java]: Thanks, addressed. I removed the separate `CountConstantRewrite`/rule type and moved the row-independent expression handling into the existing `CountLiteralRewrite`. The analyzer/rewriter now reuse `CountLiteralRewrite` before `NormalizeAggregate`, and the existing post-normalization literal rewrite remains in place.

Validated with:
- `./run-fe-ut.sh --run org.apache.doris.nereids.rules.rewrite.CountLiteralRewriteTest`
- local regression `count_constant_rewrite` via `doris-local-regression ... all -d nereids_rules_p0/count_constant_rewrite -s count_constant_rewrite`
- `git diff --check`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CountLiteralRewrite.java [hunk 2]
```diff
@@ -107,10 +114,29 @@ private boolean isCountLiteral(AggregateFunction aggFunc) {
...
+                && !arg.containsNondeterministic()
+                && !arg.containsVolatileExpression()
+                && arg.getInputSlots().isEmpty()
+                && !arg.containsType(AggregateFunction.class, SubqueryExpr.class,
+                        TableGeneratingFunction.class, WindowExpression.class);
+    }
+
     private Expression rewrite(Count count) {
         if (count.child(0).isNullLiteral()) {
             return new BigIntLiteral(0);
```
[the remark this answers is not in the document]
Comment (mrhhsg) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CountLiteralRewrite.java:136]: Thanks, addressed with a minimal targeted fix. I removed the broader `isSafeToEvaluateAboveAggregate` direction and kept the normal count-constant rewrite unchanged. For the empty-input case, `EliminateEmptyRelation` now recognizes the `CountLiteralRewrite`-generated `if(arg is null, 0, count(*))` projection above an empty aggregate and rewrites that result to literal `0`, so the original count argument is no longer evaluated when there are no input rows.

I also added regression coverage for both sides:
- empty table + invalid JSON path returns `0`
- non-empty table + the same invalid JSON path still raises `Invalid Json Path`

Validated with:
- `./run-fe-ut.sh --run org.apache.doris.nereids.rules.rewrite.CountLiteralRewriteTest`
- `doris-local-regression ... all -d nereids_rules_p0/count_constant_rewrite -s count_constant_rewrite -forceGenOut`
- `doris-local-regression ... all -d nereids_rules_p0/count_constant_rewrite -s count_constant_rewrite`
- `git diff --check`

**Comment by mrhhsg** [comment]
run buildall

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-H: <b>Total hot run time: 28971 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpch-tools
Tpch sf100 test result on commit 8fee53cb17789b4ecfc10623a47eefb37a490495, data reload: false

------ Round 1 ----------------------------------
orders	Doris	NULL	NULL	0	0	0	NULL	0	NULL	NULL	2023-12-26 18:27:23	2023-12-26 18:42:55	NULL	utf-8	NULL	NULL	
============================================
q1	17689	3999	3979	3979
q2	q3	10828	1416	804	804
q4	4686	501	361	361
q5	7555	872	606	606
q6	178	175	139	139
q7	795	858	649	649
q8	9416	1513	1678	1513
q9	5733	4495	4461	4461
q10	6768	1843	1525	1525
q11	438	277	256	256
q12	643	436	292	292
q13	18132	3339	2800	2800
q14	268	259	245	245
q15	q16	825	773	723	723
q17	924	899	1011	899
q18	6823	5854	5534	5534
q19	1328	1241	1069	1069
q20	517	407	275	275
q21	6488	2791	2526	2526
q22	444	373	315	315
Total cold run time: 100478 ms
Total hot run time: 28971 ms

----- Round 2, with runtime_filter_mode=off -----
orders	Doris	NULL	NULL	150000000	42	6422171781	NULL	22778155	NULL	NULL	2023-12-26 18:27:23	2023-12-26 18:42:55	NULL	utf-8	NULL	NULL	
============================================
q1	4765	4707	4756	4707
q2	q3	4798	4925	4321	4321
q4	2130	2198	1384	1384
q5	4411	4253	4311	4253
q6	221	176	120	120
q7	1742	1631	1417	1417
q8	2228	1944	1920	1920
q9	7365	7374	7429	7374
q10	4725	4656	4276	4276
q11	545	393	368	368
q12	733	745	539	539
q13	3017	3380	2853	2853
q14	287	284	251	251
q15	q16	696	701	619	619
q17	1283	1246	1251	1246
q18	7264	6796	6913	6796
q19	1103	1106	1139	1106
q20	2237	2254	1949	1949
q21	5312	4571	4482	4482
q22	513	446	430	430
Total cold run time: 55375 ms
Total hot run time: 50411 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-DS: <b>Total hot run time: 169914 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpcds-tools
TPC-DS sf100 test result on commit 8fee53cb17789b4ecfc10623a47eefb37a490495, data reload: false

query5	4322	652	489	489
query6	452	214	178	178
query7	4800	568	308	308
query8	369	227	229	227
query9	8779	3942	4001	3942
query10	443	312	259	259
query11	5808	2334	2200	2200
query12	167	106	100	100
query13	1277	614	434	434
query14	6464	5392	5027	5027
query14_1	4364	4364	4366	4364
query15	203	197	179	179
query16	1000	447	386	386
query17	901	695	571	571
query18	2435	480	340	340
query19	198	180	144	144
query20	109	110	108	108
query21	214	138	118	118
query22	13705	13684	14117	13684
query23	17281	16586	16257	16257
query23_1	16432	16164	16319	16164
query24	7509	1770	1295	1295
query24_1	1335	1313	1309	1309
query25	578	468	415	415
query26	1325	331	171	171
query27	2633	558	355	355
query28	4458	2091	2022	2022
query29	1108	624	511	511
query30	307	245	203	203
query31	1110	1091	981	981
query32	114	64	59	59
query33	540	327	259	259
query34	1182	1176	632	632
query35	777	789	681	681
query36	1381	1385	1150	1150
query37	155	105	96	96
query38	3202	3130	3024	3024
query39	951	924	918	918
query39_1	899	880	879	879
query40	236	129	108	108
query41	72	73	80	73
query42	96	94	93	93
query43	323	326	282	282
query44	
query45	198	192	185	185
query46	1088	1212	749	749
query47	2403	2396	2278	2278
query48	373	393	298	298
query49	648	481	379	379
query50	1044	356	257	257
query51	4335	4260	4336	4260
query52	94	94	133	94
query53	252	276	200	200
query54	293	260	199	199
query55	78	74	72	72
query56	231	234	234	234
query57	1434	1430	1333	1333
query58	246	232	199	199
query59	1527	1658	1457	1457
query60	315	251	227	227
query61	167	160	159	159
query62	706	649	598	598
query63	235	183	194	183
query64	2538	790	614	614
query65	
query66	1783	480	334	334
query67	29782	29779	29505	29505
query68	
query69	428	310	258	258
query70	946	982	962	962
query71	299	228	219	219
query72	3100	2794	2346	2346
query73	824	784	442	442
query74	5167	4926	4799	4799
query75	2666	2576	2263	2263
query76	2381	1144	757	757
query77	360	374	277	277
query78	12415	12504	12007	12007
query79	1405	1022	764	764
query80	1100	476	417	417
query81	511	291	241	241
query82	599	158	119	119
query83	353	276	249	249
query84	
query85	934	543	442	442
query86	417	301	291	291
query87	3365	3395	3176	3176
query88	3576	2706	2702	2702
query89	436	387	337	337
query90	1772	181	174	174
query91	179	162	138	138
query92	64	59	59	59
query93	1463	1372	891	891
query94	606	369	310	310
query95	641	396	365	365
query96	1127	796	353	353
query97	2710	2716	2561	2561
query98	209	213	204	204
query99	1161	1171	1043	1043
Total cold run time: 251203 ms
Total hot run time: 169914 ms
```
</details>

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Rewriter.java [hunk 3]
```diff
@@ -501,6 +502,7 @@ public class Rewriter extends AbstractBatchJobExecutor {
                     bottomUp(
                         // The later rule CHECK_PRIVILEGES which inherent from ColumnPruning only works
                         // if the aggregation node is normalized, so we need call NormalizeAggregate here
+                        new CountLiteralRewrite(),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Rewriter.java]: why add so many CountLiteralRewrite before NormalizeAggregate? NormalizeAggregate not process aggregate function's child

Comment (mrhhsg) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Rewriter.java]: Good point. I removed the extra pre-`NormalizeAggregate` placements from `Rewriter` and kept the minimal placement only in `Analyzer` before the initial `NormalizeAggregate`. `Rewriter` now keeps its original post-normalize `CountLiteralRewrite` path.

The only pre-normalize run needed for this PR is before the aggregate children are normalized into slots; the additional `Rewriter` placements were unnecessary for the minimal scope.

Validated with:
- `./build.sh --fe`
- `./run-fe-ut.sh --run org.apache.doris.nereids.rules.rewrite.CountLiteralRewriteTest`
- `./run-regression-test.sh --conf output/local-regression/regression-conf-46001.groovy --run -d nereids_rules_p0/count_constant_rewrite -s count_constant_rewrite`
- `./run-regression-test.sh --conf output/local-regression/regression-conf-46001.groovy --run -d statistics`
- `git diff --check`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CountLiteralRewrite.java [hunk 4]
```diff
@@ -107,10 +114,29 @@ private boolean isCountLiteral(AggregateFunction aggFunc) {
                 && aggFunc.child(0).isLiteral();
     }
 
+    private boolean isCountConstantExpression(AggregateFunction aggFunc) {
+        if (aggFunc.isDistinct()
+                || !(aggFunc instanceof Count)
+                || aggFunc.children().size() != 1) {
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (mrhhsg) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CountLiteralRewrite.java:122]: Addressed by excluding `Sleep` from the row-independent count rewrite (`!arg.containsType(Sleep.class)`). I also added FE unit and regression coverage to ensure `count(sleep(1))` is not rewritten to the new upper-project shape.

Validated with:
- `./build.sh --fe`
- `./run-fe-ut.sh --run org.apache.doris.nereids.rules.rewrite.CountLiteralRewriteTest`
- `./run-regression-test.sh --conf output/local-regression/regression-conf-46001.groovy --run -d nereids_rules_p0/count_constant_rewrite -s count_constant_rewrite`
- `git diff --check`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java [hunk 5]
```diff
@@ -73,6 +81,9 @@ public List<Rule> buildRules() {
                     ConnectContext.get().getStatementContext().getNextRelationId(),
                     agg.getOutput())
                 ).toRule(RuleType.ELIMINATE_AGG_ON_EMPTYRELATION),
```
[the remark this answers is not in the document]
Comment (mrhhsg) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java:75]: This obsolete compensation path was removed. The PR no longer relies on `EliminateEmptyRelation` to hide the count argument for empty input.

The current implementation documents the accepted trade-off: a deterministic row-independent count argument may be evaluated once above aggregation. Avoiding that by injecting an aggregate dependency into scalar subtrees broke functions with required constant arguments such as `date_trunc`, so this PR keeps the minimal rewrite and adds regression coverage for empty/runtime-empty value semantics with valid expressions.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java [hunk 6]
```diff
@@ -230,6 +241,48 @@ private boolean bothChildrenEmpty(LogicalJoin<?, ?> join) {
...
+            if (output.anyMatch(expr -> expr instanceof Count && ((Count) expr).isCountStar())) {
+                countStarSlots.add(output.toSlot().getExprId());
+            }
+        }
+        if (countStarSlots.isEmpty()) {
+            return null;
+        }
+
+        boolean changed = false;
+        List<NamedExpression> newProjects = new ArrayList<>(project.getProjects().size());
```
[the remark this answers is not in the document]
Comment (mrhhsg) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java]: Addressed by removing the `EliminateEmptyRelation` compensation entirely. There is no longer a broad pattern that rewrites user-authored `IF(json_extract(...) IS NULL, 0, count(*))` expressions, and nested expressions are handled by the normal projected rewrite shape.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CountLiteralRewrite.java [hunk 7]
```diff
@@ -107,10 +118,71 @@ private boolean isCountLiteral(AggregateFunction aggFunc) {
...
+        List<Expression> children = expression.children();
+        List<Expression> newChildren = Lists.newArrayListWithCapacity(children.size());
+        boolean changed = false;
+        for (Expression child : children) {
+            Expression newChild = deferConstantEvaluation(child);
+            newChildren.add(newChild);
+            changed |= newChild != child;
+        }
+
+        Expression rewritten = changed ? expression.withChildren(newChildren) : expression;
```
[the remark this answers is not in the document]
Comment (mrhhsg) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CountLiteralRewrite.java]: Addressed by removing the anti-folding wrapper. The current rewrite keeps the original scalar subtree unchanged above aggregation, so required constant arguments such as the `date_trunc` time unit remain constant. Regression coverage for `count(date_trunc('day', date '2024-01-02'))` was added.
